Library/ViewModels/CommentsViewModelTests.swift

```diff
@@ -111,365 +113,404 @@ internal final class CommentsViewModelTests: TestCase {
-        self.loadCommentsAndProjectIntoDataSourceProject
-          .assertValues([.template], "Same project is emitted again")
-        self.isCommentsLoading.assertValues([true, false], "loading ends")
-      }
-    }
+    withEnvironment(apiService: MockService(fetchCommentsEnvelopeResult: .success(envelope))) {
+      self.vm.inputs.configureWith(
+        project: nil,
+        update: nil
+      )
+      self.vm.inputs.viewDidLoad()
 
-    func testComments_WhenNoProjectOrUpdate_CommentsAreNotUpdatedInDataSource() {
       self.loadCommentsAndProjectIntoDataSourceComments.assertDidNotEmitValue()
       self.loadCommentsAndProjectIntoDataSourceProject.assertDidNotEmitValue()
+      self.isCommentsLoading.assertDidNotEmitValue()
 
-      let envelope = CommentsEnvelope.singleCommentTemplate
+      self.scheduler.advance()
 
-      withEnvironment(apiService: MockService(fetchCommentsEnvelopeResult: .success(envelope))) {
-        self.vm.inputs.configureWith(
-          project: nil,
-          update: nil
-        )
-        self.vm.inputs.viewDidLoad()
+      self.loadCommentsAndProjectIntoDataSourceComments.assertDidNotEmitValue()
+      self.loadCommentsAndProjectIntoDataSourceProject.assertDidNotEmitValue()
+      self.isCommentsLoading.assertDidNotEmitValue()
+    }
+  }
 
-        self.loadCommentsAndProjectIntoDataSourceComments.assertDidNotEmitValue()
-        self.loadCommentsAndProjectIntoDataSourceProject.assertDidNotEmitValue()
-        self.isCommentsLoading.assertDidNotEmitValue()
+  func testOptimisticComment() {
+    self.vm.inputs.configureWith(project: .template, update: nil)
 
-        self.scheduler.advance()
+    let envelope = CommentsEnvelope.singleCommentTemplate
 
-        self.loadCommentsAndProjectIntoDataSourceComments.assertDidNotEmitValue()
-        self.loadCommentsAndProjectIntoDataSourceProject.assertDidNotEmitValue()
-        self.isCommentsLoading.assertDidNotEmitValue()
-      }
+    withEnvironment(
+      apiService: MockService(fetchCommentsEnvelopeResult: .success(envelope)),
+      currentUser: .template
+    ) {
+      self.vm.inputs.viewDidLoad()
+
+      self.scheduler.advance()
+
+      let bodyText = "I just posted a comment."
+
+      self.vm.inputs.postCommentButtonTapped(bodyText)
+
+      let optimisticComment = Comment.optimisticComment(project: .template, user: .template, body: bodyText)
+
+      XCTAssertEqual(
+        self.loadCommentsAndProjectIntoDataSourceComments.values.last?.first?.body,
+        optimisticComment.body
+      )
+    }
+  }
+
+  func testPostComment() {
+    self.vm.inputs.configureWith(project: .template, update: nil)
+    self.vm.inputs.viewDidLoad()
+
+    withEnvironment(currentUser: .template) {
+      let bodyText = "Posting another comment."
+
+      self.vm.inputs.postCommentButtonTapped(bodyText)
+
+      self.postCommentSuccessful.assertValueCount(1, "Comment posted successfully")
     }
```

Comment:
Could you also test the `postCommentSubmitted` output?

Reply:
An assertion has been added for this in the `testDataSourceContainsFailableComment` function